enclave-tls/src/tls_wrappers/openssl/un_negotiate.c

```diff
@@ -0,0 +1,279 @@
+
+		/* If found then get the data */
+		if (strcmp(oid_buf, oid) == 0)
+			return 1;
+	}
+
+	return 0;
+}
+
+int openssl_extract_x509_extensions(X509 *crt, attestation_evidence_t *evidence)
+{
+	if (!strcmp(evidence->type, "sgx_epid")) {
+		int rc = find_extension_from_cert(crt, ias_response_body_oid,
+						  evidence->epid.ias_report,
+						  &evidence->epid.ias_report_len);
+		if (rc != 1)
+			return rc;
+
+		rc = find_extension_from_cert(crt, ias_root_cert_oid,
+					      evidence->epid.ias_sign_ca_cert,
+					      &evidence->epid.ias_sign_ca_cert_len);
+		if (rc != 1)
+			return rc;
+
+		rc = find_extension_from_cert(crt, ias_leaf_cert_oid, evidence->epid.ias_sign_cert,
+					      &evidence->epid.ias_sign_cert_len);
+		if (rc != 1)
+			return rc;
+
+		rc = find_extension_from_cert(crt, ias_report_signature_oid,
+					      evidence->epid.ias_report_signature,
+					      &evidence->epid.ias_report_signature_len);
+		return rc;
+	} else if (!strcmp(evidence->type, "sgx_ecdsa")) {
+		return find_extension_from_cert(crt, ecdsa_quote_oid, evidence->ecdsa.quote,
+						&evidence->ecdsa.quote_len);
+	} else if (!strcmp(evidence->type, "sgx_la")) {
+		return find_extension_from_cert(crt, la_report_oid, evidence->la.report,
+						&evidence->la.report_len);
+	}
+
+	return 1;
+}
+
+int find_extension_from_cert(X509 *cert, const char *oid, uint8_t *data, size_t *size)
+{
+	int result = 1;
+	const STACK_OF(X509_EXTENSION) * extensions;
+	int num_extensions;
+
+	/* Set a pointer to the stack of extensions (possibly NULL) */
+	if (!(extensions = X509_get0_extensions(cert))) {
+		ETLS_DEBUG("get NULL extensions from X509\n");
+		return -1;
+	}
+
+	/* Get the number of extensions (possibly zero) */
+	num_extensions = sk_X509_EXTENSION_num(extensions);
+
+	/* Find the certificate with this OID */
+	for (int i = 0; i < num_extensions; i++) {
+		X509_EXTENSION *ext;
+		ASN1_OBJECT *obj;
+		unsigned char oid_buf[128];
+
+		/* Get the i-th extension from the stack */
+		if (!(ext = sk_X509_EXTENSION_value(extensions, i)))
+			ETLS_ERR("failed to get X509 extension value\n");
+		/* Get the OID */
+		if (!(obj = X509_EXTENSION_get_object(ext)))
+			ETLS_ERR("failed to get the OID from object\n");
+
+		/* Get the string name of the OID */
+		if (!OBJ_obj2txt(oid_buf, sizeof(oid_buf), obj, 1))
+			ETLS_ERR("failed to get string name of the oid\n");
+
+		/* If found then get the data */
+		if (strcmp(oid_buf, oid) == 0) {
+			ASN1_OCTET_STRING *str;
+
+			/* Get the data from the extension */
+			if (!(str = X509_EXTENSION_get_data(ext)))
+				ETLS_ERR("failed to get data from teh extension\n");
+
+			if ((size_t)str->length > *size) {
+				*size = (size_t)str->length;
+
+				if (data)
+					ETLS_DEBUG("buffer is too small\n");
+			}
+			if (data) {
+				etls_memcpy_s(data, *size, str->data, (size_t)str->length);
+				*size = (size_t)str->length;
+				result = 1;
+				goto done;
+			}
+		}
+	}
+
+	result = -1;
+
+done:
+	return result;
+}
```

Comment:
Reorder the functions above. Currently must issue the build warning about the order.